File: problems/tier1-foundation/004-vending-machine/tests/cpp/part2_test.cpp
// Part 2 Tests — Maintenance Mode
// Tests operator maintenance mode: enter, exit, restock

#include <cassert>
#include <iostream>
#include <string>
using namespace std;

int part2_tests() {
    int passed = 0;
    int failed = 0;

    // Test 1: valid PIN enters maintenance mode
    try {
        reset();
        enterMaintenance("1234");
        assert(getState() == "Maintenance");
        cout << "PASS test_enter_maintenance_valid_pin" << endl;
        passed++;
    } catch (...) {
        cout << "FAIL test_enter_maintenance_valid_pin" << endl;
        failed++;
    }

    // Test 2: invalid PIN does not enter maintenance
    try {
        reset();
        enterMaintenance("wrong");
        assert(getState() == "Idle"); // should stay idle
        cout << "PASS test_enter_maintenance_invalid_pin" << endl;
        passed++;
    } catch (...) {
        cout << "FAIL test_enter_maintenance_invalid_pin" << endl;
        failed++;
    }

    // Test 3: in maintenance, user actions are blocked
    try {
        reset();
        enterMaintenance("1234");
        selectItem("Cola"); // should print warning, not crash
        assert(getState() == "Maintenance"); // stays in maintenance
        cout << "PASS test_user_blocked_in_maintenance" << endl;
        passed++;
    } catch (...) {
        cout << "FAIL test_user_blocked_in_maintenance" << endl;
        failed++;
    }

    // Test 4: exit maintenance returns to Idle
    try {
        reset();
        enterMaintenance("1234");
        exitMaintenance("1234");
        assert(getState() == "Idle");
        cout << "PASS test_exit_maintenance" << endl;
        passed++;
    } catch (...) {
        cout << "FAIL test_exit_maintenance" << endl;
        failed++;
    }

    // Test 5: restock only works in maintenance
    try {
        reset();
        restock("Cola", 10); // should be blocked outside maintenance
        enterMaintenance("1234");
        restock("Cola", 5); // should work in maintenance
        cout << "PASS test_restock_in_maintenance" << endl;
        passed++;
    } catch (...) {
        cout << "FAIL test_restock_in_maintenance" << endl;
        failed++;
    }

    cout << "PART2_SUMMARY " << passed << "/" << (passed + failed) << endl;
    return failed;
}

File: problems/tier1-foundation/004-vending-machine/tests/java/Part2Test.java
// Vending Machine — Part 2 Tests
import java.util.*;
import java.util.stream.*;

class Part2Test {
    static boolean testEnterMaintenanceValidPin() {
        try {
            reset();
            enterMaintenance("1234");
            boolean pass = getState() == "Maintenance";
            System.out.println((pass ? "PASS" : "FAIL") + ": testEnterMaintenanceValidPin");
            return pass;
        } catch (Exception e) {
            System.out.println("FAIL: testEnterMaintenanceValidPin (exception: " + e.getMessage() + ")");
            return false;
        }
    }

    static boolean testEnterMaintenanceInvalidPin() {
        try {
            reset();
            enterMaintenance("wrong");
            boolean pass = getState() == "Idle"); // should stay idle;
            System.out.println((pass ? "PASS" : "FAIL") + ": testEnterMaintenanceInvalidPin");
            return pass;
        } catch (Exception e) {
            System.out.println("FAIL: testEnterMaintenanceInvalidPin (exception: " + e.getMessage() + ")");
            return false;
        }
    }

    static boolean testUserBlockedInMaintenance() {
        try {
            reset();
            enterMaintenance("1234");
            selectItem("Cola"); // should print warning, not crash
            boolean pass = getState() == "Maintenance"); // stays in maintenance;
            System.out.println((pass ? "PASS" : "FAIL") + ": testUserBlockedInMaintenance");
            return pass;
        } catch (Exception e) {
            System.out.println("FAIL: testUserBlockedInMaintenance (exception: " + e.getMessage() + ")");
            return false;
        }
    }

    static boolean testExitMaintenance() {
        try {
            reset();
            enterMaintenance("1234");
            exitMaintenance("1234");
            boolean pass = getState() == "Idle";
            System.out.println((pass ? "PASS" : "FAIL") + ": testExitMaintenance");
            return pass;
        } catch (Exception e) {
            System.out.println("FAIL: testExitMaintenance (exception: " + e.getMessage() + ")");
            return false;
        }
    }

    static boolean testRestockInMaintenance() {
        try {
            reset();
            restock("Cola", 10); // should be blocked outside maintenance
            enterMaintenance("1234");
            restock("Cola", 5); // should work in maintenance
            boolean pass = true; // FIXME: no assertions found
            System.out.println((pass ? "PASS" : "FAIL") + ": testRestockInMaintenance");
            return pass;
        } catch (Exception e) {
            System.out.println("FAIL: testRestockInMaintenance (exception: " + e.getMessage() + ")");
            return false;
        }
    }

    public static int runTests() {
        int passed = 0, total = 0;
        total++; if (testEnterMaintenanceValidPin()) passed++;
        total++; if (testEnterMaintenanceInvalidPin()) passed++;
        total++; if (testUserBlockedInMaintenance()) passed++;
        total++; if (testExitMaintenance()) passed++;
        total++; if (testRestockInMaintenance()) passed++;
        System.out.println("PART2_SUMMARY " + passed + "/" + total);
        return passed;
    }
}
